File: test_regress/t/t_flag_csplit.pl
#!/usr/bin/env perl
if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2003 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

scenarios(vlt_all => 1);

while (1) {
    # Thi rule requires GNU make > 4.1 (or so, known broken in 3.81)
    #%__Slow.o: %__Slow.cpp
    #        $(OBJCACHE) $(CXX) $(CXXFLAGS) $(CPPFLAGS) $(OPT_SLOW) -c -o $@ $<
    if (make_version() < 4.1) {
        skip("Test requires GNU Make version >= 4.1");
        last;
    }

    compile(
        v_flags2 => ["--trace --output-split 1 --output-split-cfuncs 1 --exe ../$Self->{main_filename}"],
        verilator_make_gmake => 0,
        );

    # We don't use the standard test_regress rules, as want to test the rules
    # properly build
    run(logfile => "$Self->{obj_dir}/vlt_gcc.log",
        tee => $self->{verbose},
        cmd=>[$ENV{MAKE},
              "-C " . $Self->{obj_dir},
              "-f $Self->{VM_PREFIX}.mk",
              "-j 4",
              "VM_PREFIX=$Self->{VM_PREFIX}",
              "TEST_OBJ_DIR=$Self->{obj_dir}",
              "CPPFLAGS_DRIVER=-D".uc($Self->{name}),
              ($opt_verbose ? "CPPFLAGS_DRIVER2=-DTEST_VERBOSE=1" : ""),
              "OPT_FAST=-O2",
              "OPT_SLOW=-O0",
              "OPT_GLOBAL=-Os",
              ($param{make_flags}||""),
        ]);

    execute(
        check_finished => 1,
        );

    # Splitting should set VM_PARALLEL_BUILDS to 1 by default
    file_grep("$Self->{obj_dir}/$Self->{VM_PREFIX}_classes.mk", qr/VM_PARALLEL_BUILDS\s*=\s*1/);
    check_splits();
    check_no_all_file();
    check_gcc_flags("$Self->{obj_dir}/vlt_gcc.log");

    ok(1);
    last;
}
1;

sub check_splits {
    my $got1;
    my $gotSyms1;
    foreach my $file (glob("$Self->{obj_dir}/*.cpp")) {
        if ($file =~ /Syms__1/) {
            $gotSyms1 = 1;
        } elsif ($file =~ /__1/) {
            $got1 = 1;
        }
        check_cpp($file);
    }
    $got1 or error("No __1 split file found");
    $gotSyms1 or error("No Syms__1 split file found");
}

sub check_no_all_file {
    foreach my $file (glob("$Self->{obj_dir}/*.cpp")) {
        if ($file =~ qr/__ALL.cpp/) {
            error("__ALL.cpp file found: $file");
        }
    }
}

sub check_cpp {
    my $filename = shift;
    my $size = -s $filename;
    printf "  File %6d  %s\n", $size, $filename if $Self->{verbose};
    my $fh = IO::File->new("<$filename") or error("$! $filenme");
    my @funcs;
    while (defined(my $line = $fh->getline)) {
        if ($line =~ /^(void|IData)\s+(.*::.*){/) {
            my $func = $2;
            $func =~ s/\(.*$//;
            print "\tFunc $func\n" if $Self->{verbose};
            if ($func !~ /::_eval_initial_loop$/
                && $func !~ /::__Vconfigure$/
                && $func !~ /::trace$/
                && $func !~ /::traceInit$/
                && $func !~ /::traceFull$/
                && $func !~ /::final$/
                ) {
                push @funcs, $func;
            }
        }
    }
    if ($#funcs > 0) {
        error("Split had multiple functions in $filename\n\t" . join("\n\t", @funcs));
    }
}

sub check_gcc_flags {
    my $filename = shift;
    my $fh = IO::File->new("<$filename") or error("$! $filenme");
    while (defined(my $line = $fh->getline)) {
        chomp $line;
        print ":log: $line\n" if $Self->{verbose};
        if ($line =~ /$Self->{VM_PREFIX}\S*\.cpp/) {
            my $filetype = ($line =~ /Slow|Syms/) ? "slow" : "fast";
            my $opt = ($line !~ /-O2/) ? "slow" : "fast";
            print "$filetype, $opt, $line\n" if $Self->{verbose};
            if ($filetype ne $opt) {
                error("${filetype} file compiled as if was ${opt}: $line");
            }
        } elsif ($line =~ /\.cpp/ and $line !~ /-Os/) {
            error("library file not compiled with OPT_GLOBAL: $line");
        }
    }
}

File: test_regress/t/t_flag_csplit.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2024 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('vlt_all')


def check_splits():
    got1 = False
    gotSyms1 = False
    for filename in test.glob_some(test.obj_dir + "/*.cpp"):
        if re.search(r'Syms__1', filename):
            gotSyms1 = True
        elif re.search(r'__1', filename):
            got1 = True
    if not got1:
        test.error("No __1 split file found")
    if not gotSyms1:
        test.error("No Syms__1 split file found")


def check_no_all_file():
    for filename in test.glob_some(test.obj_dir + "/*.cpp"):
        if re.search(r'__ALL.cpp', filename):
            test.error("__ALL.cpp file found: " + filename)


def check_cpp(filename):
    size = os.path.getsize(filename)
    if test.verbose:
        print("  File %6d  %s\n" % (size, filename))
    funcs = []
    with open(filename, 'r', encoding="utf8") as fh:
        for line in fh:
            m = re.search(r'^(void|IData)\s+(.*::.*){', line)
            if not m:
                continue
            func = m.group(2)
            func = re.sub(r'\(.*$', '', func)
            if test.verbose:
                print("\tFunc " + func)
            if (re.search(r'(::_eval_initial_loop$', func) or re.search(r'::__Vconfigure$', func)
                    or re.search(r'::trace$', func) or re.search(r'::traceInit$', func)
                    or re.search(r'::traceFull$', func) or re.search(r'::final$', func)
                    or re.search(r'::prepareClone$', func) or re.search(r'::atClone$', func)):
                continue
            funcs.append(func)

    if len(funcs) > 0:
        test.error("Split had multiple functions in $filename\n\t" + "\n\t".join(funcs))


def check_gcc_flags(filename):
    with open(filename, 'r', encoding="utf8") as fh:
        for line in fh:
            line = line.rstrip()
            if test.verbose:
                print(":log: " + line)
            if re.search(r'' + test.vm_prefix + r'\S*\.cpp', line):
                filetype = "slow" if re.search(r'(Slow|Syms)', line) else "fast"
                opt = "fast" if re.search(r'-O2', line) else "slow"
                if test.verbose:
                    print(filetype + ", " + opt + ", " + line)
                if filetype != opt:
                    test.error(filetype + " file compiled as if was " + opt + ": " + line)
            elif re.search(r'.cpp', line) and not re.search(r'-Os', line):
                test.error("library file not compiled with OPT_GLOBAL: " + line)


# This rule requires GNU make > 4.1 (or so, known broken in 3.81)
#%__Slow.o: %__Slow.cpp
#        $(OBJCACHE) $(CXX) $(CXXFLAGS) $(CPPFLAGS) $(OPT_SLOW) -c -o $@ $<
if not test.make_version or float(test.make_version) < 4.1:
    test.skip("Test requires GNU Make version >= 4.1")

test.compile(v_flags2=["--trace-vcd",
                       "--output-split 1",
                       "--output-split-cfuncs 1",
                       "--exe",
                       "../" + test.main_filename],
             verilator_make_gmake=False)  # yapf:disable

# We don't use the standard test_regress rules, as want to test the rules
# properly build
test.run(logfile=test.obj_dir + "/vlt_gcc.log",
         tee=test.verbose,
         cmd=[os.environ["MAKE"],
              "-C " + test.obj_dir,
              "-f "+test.vm_prefix+".mk",
              "-j 4",
              "VM_PREFIX="+test.vm_prefix,
              "TEST_OBJ_DIR="+test.obj_dir,
              "CPPFLAGS_DRIVER=-D"+test.name.upper(),
              ("CPPFLAGS_DRIVER2=-DTEST_VERBOSE=1" if test.verbose else ""),
              "OPT_FAST=-O2",
              "OPT_SLOW=-O0",
              "OPT_GLOBAL=-Os",
              ])  # yapf:disable

test.execute()

# Splitting should set VM_PARALLEL_BUILDS to 1 by default
test.file_grep(test.obj_dir + "/" + test.vm_prefix + "_classes.mk", r'VM_PARALLEL_BUILDS\s*=\s*1')
check_splits()
check_no_all_file()
check_gcc_flags(test.obj_dir + "/vlt_gcc.log")

test.passes()
